### PR TITLE
Extern functions are now evaluated to undefined values at comptime in TypeOf calls.

### LINKED ISSUE #4328: support @TypeOf at global scope with function call inside it
Trying to use `lua_pop` from the lua headers I get:
```
 zig-cache/o/ZCb0oZZan0MQQOP7jNdmUMWbepVFQwtd4ZIV7svvovshAx5y8GlWK2AU6AlXbBgC/cimport.zig:921:47: error: unable to evaluate constant expression
pub inline fn lua_pop(L: var, n: var) @TypeOf(lua_settop(L, -n - 1)) {
                                              ^
```

the definition is:
```C
LUA_API void  (lua_settop) (lua_State *L, int idx);
#define lua_pop(L,n)            lua_settop(L, -(n)-1)
```

translated to:
```zig
pub extern fn lua_settop(L: ?*lua_State, idx: c_int) void;
pub inline fn lua_pop(L: var, n: var) @TypeOf(lua_settop(L, -n - 1)) {
    return lua_settop(L, -n - 1);
}
```

### PR DESCRIPTION
**Pull request by alexnask**
This lets us use `@TypeOf(extern_call(...))` like translate-c generates for macro return types.  
Closes #4328.

### CODE REVIEW

File: src/ir.cpp [hunk 1]
```diff
@@ -19707,17 +19707,25 @@ static IrInstGen *ir_analyze_fn_call(IrAnalyze *ira, IrInst* source_instr,
...
 
     if (modifier == CallModifierCompileTime) {
+        bool extern_fn_in_typeof = false;
+
         // No special handling is needed for compile time evaluation of generic functions.
         if (!fn_entry || fn_entry->body_node == nullptr) {
-            ir_add_error(ira, &fn_ref->base, buf_sprintf("unable to evaluate constant expression"));
-            return ira->codegen->invalid_inst_gen;
+            // We keep evaluating extern functions directly in TypeOfs
+            if (fn_entry && source_instr->scope->id == ScopeIdTypeOf) {
```
Comment (LemonBoy) [inline src/ir.cpp:19715]: You can heavily simplify this by using something like this:
```C
if (fn_entry && source_instr->scope->id == ScopeIdTypeOf) {
    AstNode *return_type_node = fn_proto_node->data.fn_proto.return_type;
    ZigType *specified_return_type = ir_analyze_type_expr(ira, source_instr->scope, return_type_node);
    return ir_const_undef(ira, source_instr, specified_return_type);
}
```

What's the definition of "extern" you're using here?

Comment (alexnask) [reply src/ir.cpp:19715]: I'm checking for an extern function prototype, I will use the AST node instead for more clarity.  
Updating soon.

**Comment by alexnask** [comment]
@Vexu Will do.

**Comment by Vexu** [comment]
This still doesn't fix 5305, here is a minimal repro if you want to include it in this but I don't consider it a merge blocker.
```zig
const Display = @OpaqueType();
pub const _XPrivDisplay = [*c]extern struct {
    default_screen: c_int,
   // other fields omitted ...
};
pub inline fn DefaultScreen(dpy: var) @TypeOf((if (@typeInfo(@TypeOf(dpy)) == .Pointer) @ptrCast(_XPrivDisplay, @alignCast(@alignOf(_XPrivDisplay.Child), dpy)) else if (@typeInfo(@TypeOf(dpy)) == .Int and @typeInfo(_XPrivDisplay) == .Pointer) @intToPtr(_XPrivDisplay, dpy) else @as(_XPrivDisplay, dpy)).*.default_screen) {
    return (if (@typeInfo(@TypeOf(dpy)) == .Pointer) @ptrCast(_XPrivDisplay, @alignCast(@alignOf(_XPrivDisplay.Child), dpy)) else if (@typeInfo(@TypeOf(dpy)) == .Int and @typeInfo(_XPrivDisplay) == .Pointer) @intToPtr(_XPrivDisplay, dpy) else @as(_XPrivDisplay, dpy)).*.default_screen;
}
pub inline fn DefaultRootWindow(dpy: var) @TypeOf(DefaultScreen(dpy)) {
    return DefaultScreen(dpy);
}
pub fn main() void {
    const dpy: *Display = undefined;
    _ = @TypeOf(DefaultRootWindow(dpy));
}
```

**Comment by alexnask** [comment]
@Vexu   
Sorry for the delay...  
I think doing this for any arbitrary expression requires a pretty massive rewrite that I am not comfortable undertaking right now.  
Let me rename the testcase since this does not fix 5305 so that it is mergeable, this will at least affect a good number of C imports.

**Pull request merged**